File: core/src/clj/enki_core/consent.clj
;; This file is part of Enki.
;;
;; Copyright © 2016 - 2019 Oliver Wyman Ltd.
;;
;; Enki is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;;
;; Enki is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with Enki.  If not, see <https://www.gnu.org/licenses/>
;;
(ns enki-core.consent
  (:require [enki-core.event :as e]
            [clojure.spec.alpha :as s]))

;;; Store

(def store (ref {}))

(defn last-cqs-id [subject]
  (-> (@store subject) last :hash))

(defn in-order? [subject hash]
  (= (last-cqs-id subject) hash))

(defn alter-store [subject previous f & args]
  (if (in-order? subject (second previous))
    (apply alter store update subject f args)
    :reject))

(defn append-store [subject previous data]
  (alter-store subject previous concat [(assoc data :hash (hash data))]))

(defn remove-store [subject previous f]
  (alter-store subject previous (partial remove f)))

(s/def ::subject int?)
(s/def ::object string?)
(s/def ::purposes (s/coll-of string?))
(s/def ::period keyword?)
(s/def ::places (s/coll-of keyword?))
(s/def ::hash int?)
(s/def ::previous (s/or :hash int? :root nil?))

(s/def ::give-consent-data (s/keys :req-un [::subject ::object ::purposes ::period ::places ::previous]))
(s/def ::revoke-consent-data (s/keys :req-un [::subject ::previous ::hash]))

(defmethod e/command-type :give-consent [_]
  (s/cat :command keyword? :data ::give-consent-data))

(defmethod e/command-type :revoke-consent [_]
  (s/cat :command keyword? :data ::revoke-consent-data))

(defmethod e/process :give-consent [{{:keys [subject previous] :as data} :data}]
  (let [data (dissoc data :previous)]
    (append-store subject previous data)))

(defmethod e/process :revoke-consent [{{:keys [subject previous hash]} :data}]
  (remove-store subject previous #(= (% :hash) hash)))

;;; Helpers

(defn generate-give-consent [subject data]
  (let [subject-id (hash subject)]
    [:give-consent (merge data {:subject subject-id
                                :previous (last-cqs-id subject-id)})]))

(defn generate-revoke-consent [subject id]
  (let [subject-id (hash subject)]
    [:revoke-consent {:subject subject-id
                      :previous (last-cqs-id subject-id)
                      :hash id}]))

File: bank-web-ui/server/controllers/api/consent.js
/*
   This file is part of Enki.
  
   Copyright © 2016 - 2019 Oliver Wyman Ltd.
  
   Enki is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.
  
   Enki is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
  
   You should have received a copy of the GNU General Public License
   along with Enki.  If not, see <https://www.gnu.org/licenses/>
*/

import Hydra from 'hydra-js';
import repo from '../../repository/user-data';
import logger from '../../lib/logger';

const hydra = new Hydra({
  client: {
    id: process.env.HYDRA_CLIENT_ID || 'consent-app',
    secret: process.env.HYDRA_CLIENT_SECRET || 'consent-secret',
  },
  auth: {
    tokenHost: process.env.HYDRA_URL || 'http://localhost:4444',
    authorizePath: '/oauth2/auth',
    tokenPath: '/oauth2/token',
  },
  scope: 'hydra.keys.get'
});

const internalHydraHost = process.env.HYDRA_URL || 'http://localhost:4444';
const externalHydraHost = process.env.HYDRA_EXTERNAL_URL || internalHydraHost;

const consentCheckCatcher = (res, req) => (error) => {
  logger.error('Consent confirmation failed: ' + error
    + ' from ' + req.url);
  res.status(500).send(error);
};

function verifyConsent(req, res) {
  if (!req.session.login_user) {
    res.sendStatus(403);
    return;
  }
  logger.info('Verifying consent: ' + JSON.stringify(hydra));
  try { // too bad verifyConsentChallenge() doesn't reject() properly
    hydra.verifyConsentChallenge(req.params.challenge).then(({ challenge }) => {
      res.send({challenge: challenge, csrfToken: req.csrfToken()});
    }).catch(consentCheckCatcher(res, req));
  } catch (err) {
    consentCheckCatcher(res, req)({
      error: {name: err.name, message: JSON.stringify(err.details)},
      message: 'SERVER ERROR'});
  }
}

function confirmConsent(req, res) {
  if (!req.session.login_user) {
    res.sendStatus(403);
  }

  let challenge = req.body.challenge;
  let scopes = req.body.allowed_scopes;

  // const { email, email_verified, user_id: subject, name, nickname } = user
  let subject = 'username';
  // We intentionally don't provide an onRejected callback, so all errors are
  // handled in the catch().
  repo.getUserByUsername(req.session.login_user)
    .then(user => {
      subject = user.username;

      const data = {};

      if (!Array.isArray(scopes)) { // single item
        scopes = [scopes];
      }

      // // This is the openid 'profile' scope which should include some user profile data. (optional)
      // if (scopes.indexOf('profile') >= 0) {
      //   data.name = name
      //   data.nickname = nickname
      // }

      // // This is to fulfill the openid 'email' scope which returns the user's email address. (optional)
      // if (scopes.indexOf('email') >= 0) {
      //   data.email = email
      //   data.email_verified = email_verified
      // }

      logger.info('Confirming consent: ' + JSON.stringify(hydra));
      // Make sure that the consent challenge is valid
      hydra.verifyConsentChallenge(challenge).then(({ challenge: decoded }) => {
        return hydra.generateConsentResponse(challenge, subject, scopes, {}, data)
          .then(({ consent }) => {
            decoded.redir = decoded.redir.replace(internalHydraHost, externalHydraHost); // naming convention can be different internally and externally, so correct it
            logger.info('Confirming consent (redirect): ' + JSON.stringify(decoded));
            res.redirect(decoded.redir + '&consent=' + consent);
          });
      });
    })

    .catch(consentCheckCatcher(res, req));
}

module.exports = {
  verifyConsent,
  confirmConsent
};
